AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/Dependency/Pinpoint/PinpointContext.swift

```diff
@@ -19,17 +61,132 @@ class PinpointContext {
-    lazy var sessionTracker: SessionTracker = {
-        SessionTracker(context: self)
+    lazy var sessionTracker: SessionClient = {
+        SessionClient(context: self)
     }()
     
-    init(with configuration: PinpointClient.PinpointClientConfiguration) {
-        pinpointClient = PinpointClient(config: configuration)
+    private let keychainStore: KeychainStoreBehavior
+
+    init(with configuration: PinpointContextConfiguration,
+         credentialsProvider: CredentialsProvider,
+         region: String,
+         userDefaults: UserDefaultsBehaviour = UserDefaults.standard,
+         keychainStore: KeychainStoreBehavior = KeychainStore(service: Constants.Keychain.service)) throws {
+        self.configuration = configuration
+        self.keychainStore = keychainStore
+        self.userDefaults = userDefaults
+        let pinpointConfiguration = try PinpointClient.PinpointClientConfiguration(credentialsProvider: credentialsProvider,
+                                                                                   frameworkMetadata: AmplifyAWSServiceConfiguration.frameworkMetaData(),
+                                                                                   region: region)
+        pinpointClient = PinpointClient(config: pinpointConfiguration)
     }
     
-    var legacyUniqueId: String {
-        // TODO: Implement
-        fatalError("Not yet implemented")
+    private var legacyPreferencesFilePath: String? {
+        let applicationSupportDirectoryUrls = FileManager.default.urls(for: .applicationSupportDirectory,
+                                                                      in: .userDomainMask)
+        let preferencesFileUrl = applicationSupportDirectoryUrls.first?
+            .appendingPathComponent(Constants.Preferences.mobileAnalyticsRoot)
+            .appendingPathComponent(configuration.appId)
+            .appendingPathComponent(Constants.Preferences.fileName)
+        
+        return preferencesFileUrl?.path
+    }
+    
+    private func removeLegacyPreferencesFile() {
+        guard let preferencesPath = legacyPreferencesFilePath else {
+            return
+        }
+        
+        do {
+            try FileManager.default.removeItem(atPath: preferencesPath)
+        } catch {
+            Amplify.Analytics.log.verbose("Cannot remove legacy preferences file")
+        }
+    }
+    
+    private func legacyUniqueId() -> String? {
+        guard let preferencesPath = legacyPreferencesFilePath,
+              FileManager.default.fileExists(atPath: preferencesPath),
+              let preferencesJson = try? JSONSerialization.jsonObject(with: Data(contentsOf: URL(fileURLWithPath: preferencesPath)),
+                                                                      options: .mutableContainers) as? [String: String] else {
+            return nil
+        }
+        
+        return preferencesJson[Constants.Preferences.uniqueIdKey]
+    }
+    
+    private func retrieveUniqueId() -> String {
```

Comment:
nit: let's document this method :) i.e. why we're looking in three different places etc etc

Reply:
Also, here and throughout these new types instead of accessing `Amplify.Analytics.log` we should add conformance to the `DefaultLogger` 
```swift
extension PinpointContext: DefaultLogger {}
```
and use the `log` method directly on this type
```swift
self.log.verbose("...")
```
By doing so log entries will show up with the `[PinpointContext]` prefix, making debug easier.

Reply:
> why we're looking in three different places 

I honestly have no idea, I can deduce that we've been migrating the place where we store that id through the years and kept backwards compatibility 😆 . I'll add some fancy wording at the beginning to reflect that.

> we should add conformance to the `DefaultLogger`

Nice!! I didn't know about that, but we absolutely should follow that